XML changes and Load Balancing

manishvenu opened the pull request:
Fixes #7 


- [x] This change adds two xml changes to the CESM setup related to load balancing. ROOTPE_OCN=128 to give all other parts of the model a node, and NTASKS_OCN based on making sure there are less than 800 points per core, and we use every core on a node (i.e. multiple of 128)
- [ ] **REMOVED - YAGNI**: Will also include development on how to include shared code from CRR, RM6, and RCG. I'm leaning toward a subpackage called utils ....but then we have three levels of submodules🙈

manishvenu committed:
General Cleaning & Some Function Abstraction

manishvenu committed:
Clean cesm_tools to move input.nml change to a new function

manishvenu committed:
Load Balancing

manishvenu committed:
More Changes independent of change MOMparam and setup logging

manishvenu committed:
High points of replacement that can be repalced with already written functions

manishvenu committed:
.

manishvenu committed:
Load Balancing

manishvenu committed:
Merge branch 'main' into xml_changes_load_balancing

manishvenu committed:
Rand

manishvenu committed:
Rand 2'

manishvenu commented:
Hey @mnlevy1981,

Does this PR make sense for load balancing on the ocean only case? Also, if you have the time, we made a few XML changes in lines 241 to 255. Do you have any recommendations on additional xml changes?

Thanks,
Manish V.

manishvenu commented:
@gseijo Do you have any opinions on default XML changes on line 241-255? Happy to add a few here

mnlevy1981 commented on cesm_tools.py:
```diff
@@ -221,6 +223,18 @@ def setup_cesm_explicit(
+        total_number_of_points = nx * ny
+        nodes=1
+        pts_per_processor = total_number_of_points/ float(processors_per_node)
+        while not pts_per_processor < ideal_number_of_points_per_core_ceiling:
```
How about `while pts_per_processor > ideal_number_of_points_per_core_ceiling`? We probably are okay with the edge case of `pts_per_processor == ideal_number_of_points_per_core_ceiling` and don't need to enforce it to be strictly less than.

manishvenu replied:
Sounds good!

mnlevy1981 commented on cesm_tools.py:
```diff
@@ -221,6 +223,18 @@ def setup_cesm_explicit(
+        nodes=1
+        pts_per_processor = total_number_of_points/ float(processors_per_node)
+        while not pts_per_processor < ideal_number_of_points_per_core_ceiling:
+            pts_per_processor = total_number_of_points / float(processors_per_node * nodes)
+            nodes = nodes+1
```
I think you want to set `nodes = nodes+1` before updating `pts_per_processor`. For example, if your grid has 900*128 = 115200 cells, your logic currently is:

(line 229) `pts_per_processor = 115200 / 128 = 900`
(line 230) `not 900 < 800` so enter the loop
(line 231) `nodes` is still 1, so `pts_per_processor` is still 900
(line 232) `nodes = 2`
(line 230) `not 900 < 800` so enter the loop
(line 231) `nodes` is 2, so `pts_per_processor` is 450
(line 232) `nodes = 3`
(line 230) `450 < 800` so do not enter the loop
(line 237) `nodes = 3`, so we're requesting 384 tasks for the ocean (but we'd be happy with 256 since that would end up with `pts_per_processor = 450`)

Swapping these two lines would be

(line 229) `pts_per_processor = 115200 / 128 = 900`
(line 230) `not 900 < 800` so enter the loop
(line 231) `nodes = 2`
(line 232) `nodes` is 2, so `pts_per_processor` is 450
(line 230) `450 < 800` so do not enter the loop
(line 237) `nodes = 2`, so we're requesting 256 tasks for the ocean

manishvenu replied:
Hey Mike,

Shoot that was a silly mistake. I'll clean that up.

Thanks,
Manish

manishvenu committed:
Fixes

manishvenu committed:
Standardize names

manishvenu committed:
Black Formatting

manishvenu committed:
Revert "Black Formatting"

This reverts commit b51c9cc682a11af0b2ddb63dc5c29e9ea5337f8d.

manishvenu committed:
Standardize

mnlevy1981 commented:
As for your question about the XML changes, I'm confused about why

```
        self.xmlchange(CESMPath, "MASK_GRID", str(mom_input_dir / "esmf_mesh.nc"))
        self.xmlchange(CESMPath, "OCN_GRID", str(mom_input_dir / "esmf_mesh.nc"))
        self.xmlchange(CESMPath, "ICE_GRID", str(mom_input_dir / "esmf_mesh.nc"))
```

are point to a file -- in my CESM global cases, all the `_GRID` variables are the shortnames of the grids:

```
	ATM_GRID: TL319
	GLC_GRID: null
	ICE_GRID: tx2_3v2
	LND_GRID: null
	MASK_GRID: tx2_3v2
	OCN_GRID: tx2_3v2
	ROF_GRID: JRA025
	WAV_GRID: wgx3v7
```

So we'd want (for example) `OCN_GRID=NWA12`, not a link to a mesh file. I know we use `OCN_GRID` with `ParamGen` to make sure we get the correct default value, so maybe we'd want something like `OCN_GRID=RegCaseGen` and then we can have a whole bunch of parameter defaults for regional runs that are different for global runs?

manishvenu committed:
Changed Grid XML change

manishvenu commented:
> As for your question about the XML changes, I'm confused about why
> 
> ```
>         self.xmlchange(CESMPath, "MASK_GRID", str(mom_input_dir / "esmf_mesh.nc"))
>         self.xmlchange(CESMPath, "OCN_GRID", str(mom_input_dir / "esmf_mesh.nc"))
>         self.xmlchange(CESMPath, "ICE_GRID", str(mom_input_dir / "esmf_mesh.nc"))
> ```
> 
> are point to a file -- in my CESM global cases, all the `_GRID` variables are the shortnames of the grids:
> 
> ```
> 	ATM_GRID: TL319
> 	GLC_GRID: null
> 	ICE_GRID: tx2_3v2
> 	LND_GRID: null
> 	MASK_GRID: tx2_3v2
> 	OCN_GRID: tx2_3v2
> 	ROF_GRID: JRA025
> 	WAV_GRID: wgx3v7
> ```
> 
> So we'd want (for example) `OCN_GRID=NWA12`, not a link to a mesh file. I know we use `OCN_GRID` with `ParamGen` to make sure we get the correct default value, so maybe we'd want something like `OCN_GRID=RegCaseGen` and then we can have a whole bunch of parameter defaults for regional runs that are different for global runs?

Hey @mnlevy1981,

So, I did retry running without these xml changes, and it looks like MASK_GRID and ICE_GRID doesn't change anything (which I should have known), but not setting OCN_GRID results in a bunch of missing files in the input directory (but case.submit still runs). Alper told me it’s because ParamGen works with user_nl_mom which overrides MOM_input, so it's not reading MOM_input/override correctly because OCN_GRID is set to a default (RM6 default is t232 I think). A good example is that one of the warnings was:

`Model mom missing file ocean_topog = '{$INPUT_DIR}/ocean_topog_230413.nc'` 

That make sense because in this case MOM_input has TOPO_FILE set to `bathymetry.nc` and user_nl_mom has overriden. I terminated this because I didn't think it would work with this kind of warning.


Then I set OCN_GRID=CHEESE and it works, so based on reading Alper's blog post, it may be just that if we don't use a set parameter specification we're fine and it won't override MOM_input (thought I didn't check if it actually did). A future task would be maybe to shift MOM_override changes to user_nl_mom, but maybe not for this task….unless you think it's worth figuring out right now. 

For now, I've removed MASK_GRID and ICE_GRID, and changed OCN_GRID to "RegionalCaseGen". Something random that will not make ParamGen override our changes in MOM_override.

manishvenu committed:
RegCaseGen

manishvenu commented:
Quick Status Update: Alper and I discussed switching RegCaseGen into VisualCaseGen, so more dev on this PR is paused. Will merge as is in a few hours

manishvenu committed:
Adding Logging Util

manishvenu committed:
Black and minor changes
